File: src-exe/Enigmas.hs
module Enigmas (EnigmaState(..), initialEnigmaState, renderEnigma, enigmaEventHandler, enigmas) where

import Graphics.Gloss
import Graphics.Gloss.Interface.Pure.Game

data EnigmaState = EnigmaState
  { enigmaIndex :: Int
  , enigmaQuestion :: String
  , enigmaOptions :: [String]
  , correctAnswer :: Int
  , selectedOption :: Int
  } deriving (Eq, Show)

-- Função para criar um estado inicial do Enigma usando o enigma do índice 0
initialEnigmaState :: Int -> EnigmaState
initialEnigmaState n = 
  let index = n 
  in EnigmaState  
  { enigmaIndex = index
  , enigmaQuestion = question (enigmas !! index)
  , enigmaOptions = options (enigmas !! index)
  , correctAnswer = correctAnswerIndex (enigmas !! index)
  , selectedOption = -1  -- Nenhuma opção selecionada
  }

-- Função para renderizar o enigma na tela
renderEnigma :: EnigmaState -> Picture
renderEnigma enigmaState = Pictures
  [ Translate (-500) 200 $ Scale 0.15 0.15 $ Color white $ Text (enigmaQuestion enigmaState)
  , Translate (-500) 150 $ Scale 0.15 0.15 $ Color white $ Text ("1. " ++ (enigmaOptions enigmaState !! 0))
  , Translate (-500) 100 $ Scale 0.15 0.15 $ Color white $ Text ("2. " ++ (enigmaOptions enigmaState !! 1))
  , Translate (-500) 50  $ Scale 0.15 0.15 $ Color white $ Text ("3. " ++ (enigmaOptions enigmaState !! 2))
  , Translate (-500) 0   $ Scale 0.15 0.15 $ Color white $ Text ("4. " ++ (enigmaOptions enigmaState !! 3))
  , Translate (-500) (-100) $ Scale 0.15 0.15 $ Color yellow $ Text ("Selected: " ++ show (selectedOption enigmaState + 1))
  , Translate (-500) (-150) $ Scale 0.15 0.15 $ Color yellow $ Text ("Pressione 'Enter' para confirmar sua resposta")
  ]

-- Função para manipular a seleção do enigma
enigmaEventHandler :: Event -> EnigmaState -> EnigmaState
enigmaEventHandler (EventKey (Char '1') Down _ _) enigmaState = enigmaState { selectedOption = 0 }
enigmaEventHandler (EventKey (Char '2') Down _ _) enigmaState = enigmaState { selectedOption = 1 }
enigmaEventHandler (EventKey (Char '3') Down _ _) enigmaState = enigmaState { selectedOption = 2 }
enigmaEventHandler (EventKey (Char '4') Down _ _) enigmaState = enigmaState { selectedOption = 3 }
enigmaEventHandler _ enigmaState = enigmaState

data Enigma = Enigma
  { question :: String            -- A pergunta do enigma
  , options  :: [String]          -- As opções de resposta
  , correctAnswerIndex :: Int     -- O índice da resposta correta
  }

type Enigmas = [Enigma]

enigmas :: Enigmas
enigmas =
  [ Enigma "Se você volta ao início do labirinto várias e várias vezes, então você é uma:"
      ["Pilha", "Fila", "Estrutura de repetição", "Recursão"]
      2

  , Enigma "Eu sou a seguinte sequência de números: [0, 1, 1, 2, 3, 5, 8, 13, 21, ...]. Quem sou eu?"
      ["Sequência de Fibonacci", "Sequência aritmética", "Sequência geométrica", "Série de Taylor"]
      0

  , Enigma "Você é o último a chegar aqui, mas será o primeiro a sair. Quem sou eu?"
      ["Fila", "Pilha", "Lista encadeada", "Árvore binária"]
      1

  , Enigma "Eu tenho um fim, mas eu sigo caminhos paralelos até chegar lá. Quem sou eu?"
      ["Máquina de Turing", "Autômato finito determinístico", "Autômato finito não determinístico", "Gramática livre de contexto"]
      2

  , Enigma "Você se comunica com plataformas diferentes, servindo como um intermediário entre quem usa e quem fornece. Quem sou eu?"
      ["API", "Banco de dados", "Sistema operacional", "Middleware"]
      0

  , Enigma "Se você não sabe aonde ir, mas pode verificar todos os caminhos possíveis, na pior das hipóteses, precisará voltar ao início de sua jornada. Qual técnica está sendo descrita?"
      ["Programação dinâmica", "Backtracking", "Greedy", "Busca em largura"]
      1

  , Enigma "Seu nome é uma estrutura de dados usada para análise, verificação e validação. Quem sou eu?"
      ["Grafo", "Pilha", "Árvore sintática", "Fila de prioridade"]
      2

  , Enigma "Eu sigo uma política de cache de memória na qual você é o primeiro a chegar e o primeiro a sair. Quem sou eu?"
      ["LIFO", "FIFO", "MRU", "LFU"]
      1
  ]

File: prolog/final/enigmas.pl
% Define os enigmas associados a cada labirinto
% O formato será: enigma(LabIndex, EnigmaTexto, Opções, RespostaCorreta).

enigma(1, "Se você volta ao início do labirinto repetidamente, então você é um(a): ",
        ["1.Pilha", "2.Fila", "3.Estrutura de laço", "4.Recursão"], 3).

enigma(2, "Sou a seguinte sequência de números: [0, 1, 1, 2, 3, 5, 8, 13, 21, ...]. Quem sou eu?",
        ["1.Sequência de Fibonacci", "2.Sequência aritmética", "3.Sequência geométrica", "4.Série de Taylor"], 1).

enigma(3, "Você é o último a chegar aqui, mas será o primeiro a sair. Quem é você?",
        ["1.Fila", "2.Pilha", "3.Lista encadeada", "4.Árvore binária"], 2).

enigma(4, "Tenho um fim, mas sigo caminhos paralelos até chegar lá. Quem sou eu?",
        ["1.Máquina de Turing", "2.Autômato finito determinístico", "3.Autômato finito não determinístico", "4.Gramática livre de contexto"], 3).

enigma(5, "Você se comunica com diferentes plataformas, agindo como um intermediário entre o usuário e o provedor. Quem é você?",
        ["1.API", "2.Banco de dados", "3.Sistema operacional", "4.Middleware"], 1).

enigma(6, "Você não sabe para onde ir, mas verifica todos os caminhos possíveis, possivelmente voltando ao início da sua jornada. Que técnica é você?",
        ["1.Programação dinâmica", "2.Backtracking", "3.Guloso", "4.Busca em largura"], 2).

enigma(7, "Seu nome é uma estrutura de dados usada para análise, verificação e validação. Qual é o seu nome?",
        ["1.Grafo", "2.Pilha", "3.Árvore sintática", "4.Fila de prioridade"], 3).

enigma(8, "Eu sigo uma política de memória cache onde você é o primeiro a chegar e o primeiro a sair. Quem sou eu?",
        ["1.LIFO", "2.FIFO", "3.MRU", "4.LFU"], 2).

enigma(9, "Eu sou o meio, com base no início, ao final de um caminho. Às vezes, me chamam de 'research'. Quem sou eu?",
        ["1.Análise", "2.Metodologia", "3.Referencial teórico", "4.Referências"], 2).

% Função para obter o enigma associado ao labirinto
initial_enigma_state(LabIndex, EnigmaState) :-
    enigma(LabIndex, Texto, Opcoes, RespostaCorreta),
    EnigmaState = enigma(Texto, Opcoes, RespostaCorreta).

% Função para exibir o enigma
render_enigma(enigma(Texto, Opcoes, _)) :-
    write(Texto), nl,
    write('Opções:'), nl,
    forall(member(Op, Opcoes), (write(Op), nl)).

% Função para verificar se a opção escolhida é válida
valid_input(OptionNumber, enigma(_, _, _)) :-
    OptionNumber >= 1,
    OptionNumber =< 4.

% Função para lidar com a resposta do enigma
enigma_event_handler(OptionNumber, EnigmaState, UpdatedState) :-
    valid_input(OptionNumber, EnigmaState),
    UpdatedState = EnigmaState.

% Função para verificar se a opção escolhida é a correta
check_answer(OptionNumber, enigma(_, _, RespostaCorreta), 'Resposta correta! Você pode prosseguir.') :-
    OptionNumber = RespostaCorreta.

check_answer(OptionNumber, enigma(_, _, RespostaCorreta), 'Resposta incorreta. O jogo acabou.') :-
    OptionNumber \= RespostaCorreta.
